### PR TITLE
832 miovision readme updates

### LINKED ISSUE #636: Miovision: reorder documentation
In reviewing #591 I was struck that the sequence of information is logical, but is not prioritized by what an analyst might need. My initial thoughts are:

1. What is this?
2. How do I access it?
3. How do I use it?
4. How was it aggregated?
5. Data Ops: how does the pipeline work?
6. Data Ops: something went wrong predictably, how do I fix it?
7. Data Ops: something went wrong that we can't fix, who should I contact?

### LINKED ISSUE #832: Miovision Readme updates
The Miovision readmes are missing some necessary edits to reflect new state since #822.
To update:  

- [x] [Repulling data](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/README.md#4-repulling-data)
- [x] [Backfill/Aggregate new intersection data](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/update_intersections/Readme.md#backfillaggregate-new-intersection-data)
- [x] [Adding many intersections](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/update_intersections/Readme.md#adding-many-intersections)
- [x] [Volumes](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/README.md#volumes) (remove mention of triggers) 
- [x] [How to run the api](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#how-to-run-the-api)
- [x] [PostgreSQL Functions](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#postgresql-functions)
- [x] [How the api works](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#how-the-api-works)
- [x] [Miovision Pull](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#miovision_pull)
- [x] Basic description
- [ ] Improved diagram of aggregate tables

### LINKED ISSUE #858: Basic/Introductory documentation needed for Miovision data
Someone asked me for some basic information on our Miovision data, and there really doesn't seem to be anything to show them that doesn't get way too technical way too fast. 
https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/README.md

I'd like to create an introductory readme that just contains the basics.
* what are these cameras?
* what do they count?
* when?
* how?
* where are they?
* we are adding more!
* etc

The technical stuff then can be buried a little deeper, and I think would be better in linked readme's, not the main one. Basically, we need better information hierarchy here - general to specific.

### PR DESCRIPTION
**Pull request by gabrielwol**
## What this pull request accomplishes:
Many Miovision readme updates including:
- [Repulling data](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/README.md#4-repulling-data)
- [Backfill/Aggregate new intersection data](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/update_intersections/Readme.md#backfillaggregate-new-intersection-data)
- [Adding many intersections](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/update_intersections/Readme.md#adding-many-intersections)
- [Volumes](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/README.md#volumes) (remove mention of triggers)
- [How to run the api](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#how-to-run-the-api)
- [PostgreSQL Functions](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#postgresql-functions)
- [How the api works](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#how-the-api-works)
- [Miovision Pull](https://github.com/CityofToronto/bdit_data-sources/blob/master/volumes/miovision/api/readme.md#miovision_pull)
- Basic description of Miovision, restructure main Readme
- Mermaid diagrams: API script, miovision_api aggregations
- refreshed geojson file

## Issue(s) this solves:
- Closes #832, Closes #858, Closes #636

## What, in particular, needs to reviewed:
Would prefer to merge these many improvements sooner rather than later - there will always be more work to be done!  

## What needs to be done by a sysadmin after this PR is merged

Update sql comments: 

```sql
COMMENT ON FUNCTION miovision_api.aggregate_15_min(date, date, integer []) IS '''Aggregates data from `miovision_api.volumes_15min_mvt` (turning movements counts/TMC) into `miovision_api.volumes_15min` (automatic traffic recorder /ATR). Also updates `miovision_api.volumes_mvt_atr_xover` and `miovision_api.volumes_15min_mvt.processed` column. Takes an optional intersection array parameter to aggregate only specific intersections. Use `clear_volumes_15min()` to remove existing values before summarizing.''';
COMMENT ON FUNCTION miovision_api.aggregate_15_min_mvt(date, date, integer []) IS '''Aggregates valid movements from `miovision_api.volumes` in to `miovision_api.volumes_15min_mvt` as 15 minute turning movement counts (TMC) bins and fills in gaps with 0-volume bins. Also updates foreign key in `miovision_api.volumes`. Takes an optional intersection array parameter to aggregate only specific intersections. Use `clear_15_min_mvt()` to remove existing values before summarizing.''';
COMMENT ON FUNCTION miovision_api.api_log(date, date, integer []) IS '''Logs inserts from the api to miovision_api.volumes via the `miovision_api.api_log` table. Takes an optional intersection array parameter to aggregate only specific intersections. Use `clear_api_log()` to remove existing values before summarizing.''';
COMMENT ON FUNCTION miovision_api.get_report_dates(timestamp, timestamp, integer []) IS '''Logs the intersections/classes/dates added to `miovision_api.volumes_15min` to `miovision_api.report_dates`. Takes an optional intersection array parameter to aggregate only specific intersections. Use `clear_report_dates()` to remove existing values before summarizing.''';
COMMENT ON FUNCTION miovision_api.clear_15_min_mvt(timestamp, timestamp, integer []) IS '''Clears data from `miovision_api.volumes_15min_mvt` in order to facilitate re-pulling. `intersections` param defaults to all intersections.''';
COMMENT ON FUNCTION miovision_api.clear_api_log(date, date, integer []) IS '''Clears data from `miovision_api.api_log` in order to facilitate re-pulling. `intersections` param defaults to all intersections.''';
COMMENT ON FUNCTION miovision_api.clear_report_dates(date, date, integer []) IS '''Clears data from `miovision_api.report_dates` in order to facilitate re-pulling. `intersections` param defaults to all intersections.''';
COMMENT ON FUNCTION miovision_api.clear_volumes(timestamp, timestamp, integer []) IS '''Clears data from `miovision_api.volumes` in order to facilitate re-pulling. `intersections` param defaults to all intersections.''';
COMMENT ON FUNCTION miovision_api.clear_volumes_15min(timestamp, timestamp, integer []) IS '''Clears data from `miovision_api.volumes_15min` in order to facilitate re-pulling. `intersections` param defaults to all intersections.''';
COMMENT ON FUNCTION miovision_api.find_invalid_movements(timestamp, timestamp) IS '''Used exclusively within `intersection_tmc.py` `insert_data` function to raise notice in the logs about invalid movements.''';
COMMENT ON FUNCTION miovision_api.get_intersections_uids(integer []) IS '''Returns all intersection_uids if optional `intersections` param is omitted, otherwise returns only the intersection_uids provided as an integer array to intersections param. Used in `miovision_api.clear_*` functions. Example usage: `SELECT miovision_api.get_intersections_uids() --returns all intersection_uids` or `SELECT miovision_api.get_intersections_uids(ARRAY[1,2,3]::integer[]) --returns only {1,2,3}`''';
COMMENT ON FUNCTION miovision_api.create_mm_nested_volumes_partitions(text, integer, integer) IS '''Create a new month partition under the parent year table `base_table`. Only to be used for miovision_api `volumes_15min` and `volumes_15min_mvt` tables. Example: `SELECT miovision_api.create_yyyy_volumes_partition(''volumes_15min'', 2023)`''';
COMMENT ON FUNCTION miovision_api.create_yyyy_volumes_15min_partition(text, integer) IS '''Create a new year partition under the parent table `base_table`. Only to be used for miovision_api `volumes_15min` and `volumes_15min_mvt` tables. Example: `SELECT miovision_api.create_yyyy_volumes_partition(''volumes_15min'', 2023)`''';
COMMENT ON FUNCTION miovision_api.create_yyyy_volumes_partition(text, integer, text) IS '''Create a new year partition under the parent table `base_table`. Only to be used for miovision_api `volumes` table. Use parameter `datetime_col` to specify the partitioning timestamp column, ie. `datetime_bin`. Example: `SELECT miovision_api.create_yyyy_volumes_partition(''volumes'', 2023, ''datetime_bin'')`''';
COMMENT ON FUNCTION miovision_api.determine_working_machine(date, date) IS '''Function no longer in use. Previously used in `check_miovision` DAG to determine if any cameras had gaps larger than 4 hours. See: `miovision_check` DAG `check_gaps` task for new implementation.''';
COMMENT ON FUNCTION miovision_api.missing_dates(date) IS '''Function no longer in use. Previously used to log dates with missing data to `miovision_api.missing_dates`.''';
COMMENT ON FUNCTION miovision_api.identify_zero_counts(date, integer []) IS 'Identifies intersection / classification (only classification_uid 1,2,6,10) combos with zero volumes for the start_date called. Inserts or updates anomaly into anomalous_range table unless an existing, overlapping, manual entery exists.';
```

### CODE REVIEW

File: volumes/miovision/sql/README.md [hunk 1]
```diff
@@ -1,3 +1,549 @@
...
+| [`api_log(start_date date, end_date date, intersections integer[])`](function/function-api_log.sql) | Logs inserts from the api to miovision_api.volumes via the `miovision_api.api_log` table. Takes an optional intersection array parameter to aggregate only specific intersections. Use `clear_api_log()` to remove existing values before summarizing. |
+| [`get_report_dates(start_date timestamp, end_date timestamp, intersections integer[])`](function/function-get_report_dates.sql) | Logs the intersections/classes/dates added to `miovision_api.volumes_15min` to `miovision_api.report_dates`. Takes an optional intersection array parameter to aggregate only specific intersections. Use `clear_report_dates()` to remove existing values before summarizing. |
+| [`find_gaps(start_date date, end_date date)`](function/function-find_gaps.sql) | Find unacceptable gaps and insert into table `miovision_api.unacceptable_gaps`. |
```
Comment (chmnata) [inline volumes/miovision/sql/README.md:455]: and the new find zero function

File: volumes/miovision/sql/README.md [hunk 2]
```diff
@@ -1,3 +1,549 @@
...
+
+1. Within `find_gaps` function, the `gapsize_lookup_insert` function is run to populate the table `gapsize_lookup` with acceptable gap sizes based on avg hourly volumes from a 60 day rolling lookback. 
+  - The set of acceptable `gap_tolerance` implemented is based on an investigation stated in this [notebook](../dev_notebooks/volume_vs_gaps.ipynb). 
+2. Then, the function [`miovision_api.find_gaps`](function-find_gaps.sql) identifies all gaps of data in the table `miovision_api.volumes` and check if they are within the acceptable range of gap sizes or not based on the information from the `gapsize_lookup` table above. Sensors with no data are not identified - these will instead be included in `anomalous_ranges` table with PR [#777](https://github.com/CityofToronto/bdit_data-sources/pull/777).
```
Comment (chmnata) [inline volumes/miovision/sql/README.md:498]: can change the PR to the function meow

File: volumes/miovision/sql/README.md [hunk 3]
```diff
@@ -1,3 +1,549 @@
...
+        SELECT 1 FROM ref.holiday WHERE holiday.dt = volumes.datetime_bin::date
+    )
+    AND etc.
+```
+
+### Identifying new anomalies
+The `anomalous_ranges` table is populated in different ways:
+  - manually by `miovision_data_detectives` after visual inspection or prompting from an Airflow alert
+  - automatically by a [daily script](function/function-identify-zero-counts.sql) which identifys zero volume days by intersection/classification.
+    - There is an intention to eventually flag more unusual volumes automagically (see [Issue #630](https://github.com/CityofToronto/bdit_data-sources/issues/630)).
```
Comment (chmnata) [inline volumes/miovision/sql/README.md:548]: can update this as well

File: volumes/miovision/sql/table/create-mv-mio_cent.sql [hunk 4]
```diff
@@ -107,7 +107,7 @@ SELECT
     dl.intersection_uid AS data_int,
     dl.leg AS data_leg
 FROM data_legs AS dl
-FULL JOIN miovision_api.centerline_miovision AS cm ON cm.intersection_uid = dl.intersection_uid AND cm.leg = dl.leg -- full join means I can see what's in each pile
+FULL JOIN miovision_api.centreline_miovision AS cm ON cm.intersection_uid = dl.intersection_uid AND cm.leg = dl.leg -- full join means I can see what's in each pile
```
Comment (chmnata) [inline volumes/miovision/sql/table/create-mv-mio_cent.sql:110]: probs not in the scope of this PR but we should be referencing the gis_core centreline and intersection table meow

File: volumes/miovision/update_intersections/Readme.md [hunk 5]
```diff
@@ -48,7 +35,7 @@ Adding intersections is not as simple as removing an intersection. We will first
 Look at the table [`miovision_api.intersections`](../README.md#intersections) to see what information about the new intersections is needed to update the table. The steps needed to find details such as id, coordinates, px, int_id, geom, which leg_restricted etc are descrived below. Once everything is done, have a member of `miovision_admins` do an INSERT INTO this table to include the new intersections.
```
Comment (chmnata) [inline volumes/miovision/update_intersections/Readme.md:35]: described typo

File: volumes/miovision/update_intersections/Readme.md [hunk 6]
```diff
@@ -59,13 +46,30 @@ Look at the table [`miovision_api.intersections`](../README.md#intersections) to
 4. **px**  
     `px` can be found by searching the intersection name (location) in ITS Central (https://itscentral.corp.toronto.ca/) and finding the corresponding intersection id (PX####). `px` id can be used to look up the rest of the information (`street_main`, `street_cross`, `geom`, `lat`, `lng` and `int_id`) from table `gis.traffic_signal` as in the query below. Note that `px` is a zero padded text format in `gis.traffic_signal`, but stored as an integer in `miovision_api.intersections`.
```
Comment (chmnata) [inline volumes/miovision/update_intersections/Readme.md:47]: could say that px is the id for traffic signal

File: volumes/miovision/update_intersections/Readme.md [hunk 7]
```diff
@@ -59,13 +46,30 @@ Look at the table [`miovision_api.intersections`](../README.md#intersections) to
...
+
 	<img src="image-1.png" alt="Identifying miovision `px` using ITS Central" width="600"/>
 		
 4. **Restricted legs**  
-    In order to find out which leg of that intersection is restricted, go to Google Map to find out the direction of traffic.
+    In order to find out which leg of that intersection is restricted (no cars approaching from that leg), go to Google Map to find out the direction of traffic.
 
 6. **Insert statement**  
-    Prepare an insert statement for the new intersection(s). Alternatively [this](#adding-many-intersections) section contains a python snippet you can use to do the same, which may be helpful for adding a large number of intersections. 
+    Prepare an insert statement for the new intersection(s). Alternatively [this](#adding-many-intersections) section contains a python snippet you can use to do the same, which may be helpful for adding a large number of intersections or if you want to insert with the help of the miovision_bot due to permissions issues.
```
Comment (chmnata) [inline volumes/miovision/update_intersections/Readme.md:72]: sneaky bot

File: volumes/miovision/update_intersections/Readme.md [hunk 8]
```diff
@@ -221,33 +252,10 @@ We need to find out all valid movements for the new intersections from the data
...
-	import psycopg2
-	from psycopg2.extras import execute_values
-
-	import configparser
-	import pathlib
-
-	# Insert code to read configuration settings.
-	postgres_settings = {your_postgres_config}
+5. **Update geojson**
+	Update the [geojson intersections file](../geojson/mio_intersections.geojson) by exporting to file from QGIS with `GeoJSON` format. This geojson file is helpful as a publically accessible record of our Miovision intersections.
```
Comment (chmnata) [inline volumes/miovision/update_intersections/Readme.md:256]: one more line between 255 and 256

File: volumes/miovision/update_intersections/Readme.md [hunk 9]
```diff
@@ -221,33 +252,10 @@ We need to find out all valid movements for the new intersections from the data
...
-	with psycopg2.connect(**postgres_settings) as conn:
-		with conn.cursor() as cur:
-			insert_data = """INSERT INTO miovision_api.intersection_movements(intersection_uid, classification_uid, leg, movement_uid) VALUES %s"""
-			execute_values(cur, insert_data, df_list)
-			if conn.notices != []:
-				print(conn.notices)
-	```
 ## Update `miovision_api.centreline_miovision`
 
 [`miovision_api.centreline_miovision`](../README.md) links Miovision intersection legs to `gis.centreline` street segments. The script used to create this table is [here](../sql/table/create-mv-mio_cent.sql).
```
Comment (chmnata) [inline volumes/miovision/update_intersections/Readme.md:261]: gis_core

**Review by chmnata — CHANGES_REQUESTED** [review]
Thanks @gabrielwol, this is so much clearer and organized, bet the report card will look better after this :meow_ez:

Just have some minor comments.

**Comment by Nate-Wessel** [comment]
@gabrielwol one more suggestion for the introductory readme: can we add a couple pictures? I think it would be helpful to just show what the cameras actually look like out in the world, and perhaps also an example or two of what they see through their fisheye lenses and how that's split up and configured.

**Comment by jwrcoleman** [comment]
Love to see this! I'd not bother mentioning the CSV thing in here, think it'll just cause confusion for most. Would love to see a limitations section in here where we reference that bicycle counts are still going through a validation/improvement process and whatever other limitations are worth highlighting.

**Comment by radumas** [comment]
> One more thing, I wonder if the folder /csv is still relevant? @radumas

the one with the King Street exceptions table, or the folder under `sql`?

**Comment by chmnata** [comment]
> > One more thing, I wonder if the folder /csv is still relevant? @radumas
> 
> the one with the King Street exceptions table, or the folder under `sql`?

Ah didnt know that was King Street's exception table, but yes that one.

File: volumes/miovision/sql/README.md (hunk 3 above)
Comment (gabrielwol) [reply volumes/miovision/sql/README.md:548]: This comment was referring to other auto-flagging ideas brought up in that issue!

**Comment by gabrielwol** [comment]
Add these limitations to main readme: 

> The data quality is evolving, so refer to the below general rule for dates at which different streams became reliable enough:
> September 2019 for bikes,
> June 2019 for pedestrians
> January 2019 for cars

File: volumes/miovision/update_intersections/Readme.md (hunk 9 above)
Comment (gabrielwol) [reply volumes/miovision/update_intersections/Readme.md]: I added a note in the centreline_miovision section of 'sql' readme that this table needs to be updated to refer to gis_core.

**Review by chmnata — APPROVED** [review]
Thanks @gabrielwol, great work on improving the doc and all the 🧜🏼‍♀️ !

**Pull request merged**